### ISSUE TITLE
It seems we created the same middleware

### ISSUE THREAD

**Issue by baseplate-admin** [adamchainz/django-minify-html#5]
Hi there again,

I created this [repo](https://github.com/baseplate-admin/django_strip_whitespace) [some time ago](https://github.com/baseplate-admin/python_strip_whitespace/tree/6839fe33a79587a588e099f3fd140a928471ef40). It also has [![Downloads](https://pepy.tech/badge/django-strip-whitespace/month)](https://pepy.tech/project/django-strip-whitespace)  and [![Downloads](https://pepy.tech/badge/django-strip-whitespace)](https://pepy.tech/project/django-strip-whitespace) 
## Which also has some traffic :

![python](https://user-images.githubusercontent.com/61817579/147600996-4b66d79e-effb-4f53-a752-0da9eb74abc8.PNG)

 It essentially does what you are doing here (+ more). But there are some problems with this approach. 

As discussed [here ](https://www.reddit.com/r/django/comments/qyqxht/comment/hlj3rzb/?utm_source=share&utm_medium=web2x&context=3) your [approach ](https://github.com/adamchainz/django-minify-html/blob/4fa88008bb0ef1eca898f6247ee2dfb557e3b41b/src/django_minify_html/middleware.py#L14) is inefficient. ( Decoding GZIP is useless ). Also your approach doesn't work for special HTML tags like ( [X-Init](https://alpinejs.dev/directives/init), [X-Effect](https://alpinejs.dev/directives/effect), [X-On](https://alpinejs.dev/directives/on) | Used by [AlpineJS](https://alpinejs.dev/) which is used in many Django project ). 

My [approach](https://raw.githubusercontent.com/baseplate-admin/python_strip_whitespace/main/python_strip_whitespace/functions/regex_patterns/alpinejs.py) [here ](https://raw.githubusercontent.com/baseplate-admin/python_strip_whitespace/main/python_strip_whitespace/html/line_break.py) fixes some of the issues.
Also to allow the usage in [other applications](https://github.com/baseplate-admin/flask_strip_whitespace) I decided to split the main repo in another [project](https://github.com/baseplate-admin/python_strip_whitespace). 

Also 2  projects for essentially one end goal is kinda useless. Because we will essentially rewrite same thing without adding any new features. Is there any chance of merging our repos together or focusing on [one repo](https://github.com/baseplate-admin/python_strip_whitespace) ( which will essentially allow this module to be used in every python application ) ?

Currently [my repo](https://github.com/baseplate-admin/python_strip_whitespace) has an [issue ](https://github.com/baseplate-admin/python_strip_whitespace/issues/1), which if fixed will be a feature complete library. I also have an idea of using [esbuild](https://github.com/evanw/esbuild) with [PyO3](https://github.com/PyO3/pyo3/) ( [which minify-html uses](https://github.com/wilsonzlin/esbuild-rs) and with which i will replace [minify-html](https://github.com/wilsonzlin/minify-html) ) to make [the module](https://github.com/baseplate-admin/django_strip_whitespace) even faster. I will also add full support for [petite-vue](https://github.com/vuejs/petite-vue) in the next year. But for now the development  is a bit slow as I am still a beginner in rust.

**Comment by adamchainz** [comment]
Hello again!

I don't believe we're *quite* rebuilding the same thing. I intended to only minify HTML. Your package takes a different approach - it's trying to do a lot more.

> As discussed here your approach is inefficient. ( Decoding GZIP is useless ). 

Please can you elaborate? My package is not decoding gzip at the point you linked. The `decode()` call converts `bytes` to `str`.

> Also your approach doesn't work for special HTML tags like ( X-Init, X-Effect, X-On | Used by AlpineJS which is used in many Django project ).

Can you explain what you mean by "doesn't work"? The attributes are valid HTML attributes, and minify-html seems to process them just fine when I run the examples:

```
In [1]: import minify_html

In [2]: minify_html.minify('''<div x-init="console.log('I\'m being initialized!')"></div>''')
Out[2]: '<div x-init="console.log(\'I\'m being initialized!\')"></div>'
```

Any problem that minify-html has with them is something I'd consider fixing upstream in minify-html, rather than patch it up when wrapping it.

I briefly looked at your package's source code, and you're using a lot of regexes. I'm afraid that approach is flawed. You cannot parse HTML with a regex - it's too complicated. It's not a *regular language*. There's even [this Stack Overflow meme answer](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454) that says if you try, you'll summon a pony.

To parse HTML, you need to use a real HTML parser like Python's `html.parser` (which is even not *that* good, since it doesn't handle many quirks or parts of the actual HTML spec). [Here's a post](https://adamj.eu/tech/2021/09/23/truncating-my-blog-posts-with-html-parser/) that I wrote on using it for my blog.

The need for a real parser is why I'm excited to use minify-html (and *o*nly minify-html). It uses a very fast HTML parser.

A secondary problem I think your package has is that you only allow *one* compression algorithm to be active. That's not how HTTP encoding is intended to work. You should inspect the `Accept-Encoding` header to determine what compression algorithm the client supports, and use the best one available. This is what [django-compression-middleware](https://pypi.org/project/django-compression-middleware/) does. I also think the compression middleware is best *separate* from the HTML minification middleware. Compression can be applied to all responses, not just HTML ones.

Thanks for writing. I hope this has given you some knowledge and ideas!

**Comment by baseplate-admin** [comment]
> I don't believe we're quite rebuilding the same thing. I intended to only minify HTML. Your package takes a different approach     it's trying to do a lot more.

Correct. But end goal is to have the same minified-html.

> Can you explain what you mean by "doesn't work"? The attributes are valid HTML attributes, and minify-html seems to process them just fine when I run the examples:

Take a look at this [repo](https://github.com/baseplate-admin/Blog/blob/b3440f12f2d97eaa0a276fcfff92ec96f6d7d448/core/templates/components/navbar.html#L1)

I ran this with:
```python
import minify_html

x = minify_html.minify(open('index.html').read(),minify_css=True,minify_js=True)

print(x)
```

### Result :
```html
{% load static %} <link href="{% static 'core/css/navbar/navbar.css' %}"rel=stylesheet><nav aria-label="main navigation"class="navbar container"x-data="{ 
            open: false,
            button_turned: false,
        }"role=navigation x-cloak><div x-init="
            () => {
                switch ( $store.is_mobile ) {
                    case true : {
                        open = false;
                        button_turned = true;
                        break;
                    }
                    case false : {
                        open = true;
                        break;
                    }
                }
            }
        "class=navbar-brand><a class="navbar-item is-clickable"href=https://bulma.io> <img src="{% static 'core/images/logo.png' %}"height=28 width=112> </a><a :class="open ? 'is-active' : ''"@click="open = !open"class="navbar-burger is-clickable"role="button "aria-expanded=false aria-label=menu data-target=navbarBasicExample> <span aria-hidden=true></span> <span aria-hidden=true></span> <span aria-hidden=true></span> </a></div><div class="navbar-menu is-active"id=mainNavbar x-collapse x-show=open><div x-show=open x-transition x-transition.delay.50ms><div :class="$store.is_mobile ? 'is-flex is-flex-direction-column' : ''"class="navbar-start mb-3"><button @click.prevent="
                        () => {
                            window.location = '/'
                        }
                    "@mouseenter="
                        () => {
                            anime({
                                targets: '.animejs__home__icon',
                                color: '#e50000',
                            })
                        }
                    "@mouseleave="
                        () => {
                            anime({
                                targets: '.animejs__home__icon',
                                color: '#d9d9d9',
                            })
                        }
                    "class="navbar-item button is-ghost is-rounded ml-1 is-unselectable"><ion-icon class=animejs__home__icon name=home-sharp></ion-icon> <p>Home</p></button><button @click.prevent="
                        () => {
                            window.location = '/blog/'
                        }
                    "@mouseenter="
                        () => {
                            anime({
                                targets: '.animejs__school__icon',
                                color: '#e50000',
                            })
                        }
                    "@mouseleave="
                        () => {
                            anime({
                                targets: '.animejs__school__icon',
                                color: '#d9d9d9',
                            })
                        }
                    "class="navbar-item button is-ghost is-rounded ml-2 is-unselectable"x-init="
                        () => {
                            const string = '{{ request.path }}';

                            if (string.includes('blog') ) {
                                $el.classList.add('hover')
                            }
                        }
                    "><ion-icon class=animejs__school__icon name=school-sharp></ion-icon> <p>Blog</p></button><button @mouseenter="
                        () => {
                            anime({
                                targets: '.animejs__projects__icon',
                                color: '#e50000',
                            })
                        }
                    "@mouseleave="
                        () => {
                            anime({
                                targets: '.animejs__projects__icon',
                                color: '#d9d9d9',
                            })
                        }
                    "class="navbar-item button is-ghost is-rounded ml-2 is-unselectable"x-init="
                            () => {
                                if ('{{ request.path }}' === '/projects/') {
                                    $el.classList.add('hover')
                                }
                            }
                        "><ion-icon class=animejs__projects__icon name=extension-puzzle-sharp></ion-icon> <p>Projects</p></button></div></div><div class=navbar-end><div :class="$store.is_mobile ? 'has-text-centered' : ''"class=navbar-item><a :class="$store.is_mobile ? 'button': ''"@mouseenter="
                        () => {
                            anime({
                                targets: '.animejs__logo__facebook',
                                color: '#4169e1',
                            });
                            anime({
                                targets: '.animejs__facebook__button',
                                scale: 1.3
                            });
                        }
                    "@mouseleave="
                        () => {
                            anime({
                                targets: '.animejs__logo__facebook',
                                    color: 'hsl(0, 0%, 80%)',
                            });
                            anime({
                                targets: '.animejs__facebook__button',
                                scale: 1
                            });
                        }
                    "class="is-rounded is-dark animejs__facebook__button is-clickable"href="{{ settings.site_settings.SocialMediaSettings.facebook }}"x-effect="
                        () => {
                            switch (button_turned) {
                                case true : {
                                    anime({
                                        targets: '.animejs__facebook__button',
                                        translateX: 0,
                                        easing: 'easeOutSine',
                                        duration: 150,
                                        opacity: 1 ,
                                        scale: 1,
                                    });
                                    break;
                                };
                                case false: {
                                    anime({
                                        targets: '.animejs__facebook__button',
                                        translateX: 40 * 2,
                                        easing: 'easeOutSine',
                                        duration: 150,
                                        opacity: 0,
                                        scale: 0.2,
                                    });
                                    break;
                                };
                            }
                        }
                    "x-init="
                        () => {
                            tippy('.animejs__facebook__button', {
                                content: 'Facebook',
                            });
                        }
                    "> <ion-icon :class="$store.is_mobile ? 'is-position-absolute' : ''"x-init="
                            () => {
                                anime({
                                    targets: '.animejs__logo__facebook',
                                    color: 'hsl(0, 0%, 80%)',
                                    scale: 0.6,
                                });
                            }
                        "class=animejs__logo__facebook name=logo-facebook style=width:100%;height:100%></ion-icon> </a><a :class="$store.is_mobile ? 'button ': ''"@mouseenter="
                            () => {
                                anime({
                                    targets: '.animejs__logo__github',
                                    color: 'hsl(0, 0%, 100%)',
                                });
                                anime({
                                    targets: '.animejs__github__button',
                                    scale: 1.3
                                });
                            }
                        "@mouseleave="
                        () => {
                            anime({
                                targets: '.animejs__logo__github',
                                color: 'hsl(0, 0%, 80%)',
                            });
                            anime({
                                targets: '.animejs__github__button',
                                scale: 1
                            });
                        }
                    "class="is-rounded is-dark animejs__github__button is-clickable"href="{{ settings.site_settings.SocialMediaSettings.github }} "x-effect="
                        () => {
                            switch ( button_turned ) {
                                case true : {
                                    anime({
                                        targets: '.animejs__github__button',
                                        translateX: 0,
                                        easing: 'easeOutSine',
                                        duration: 150,
                                        opacity: 1 ,
                                        scale: 1,
                                    });
                                    break;
                                };
                                case false : {
                                    anime({
                                        targets: '.animejs__github__button',
                                        translateX: 40,
                                        easing: 'easeOutSine',
                                        duration: 150,
                                        opacity: 0,
                                        scale: 0.2,
                                    });
                                    break;
                                };
                            }
                        }
                    "x-init="
                        () => {
                            tippy('.animejs__github__button', {
                                content: 'Github',
                            });
                        }
                    "> <ion-icon :class="$store.is_mobile ? 'is-position-absolute' : ''"x-init="
                            () => {
                                anime({
                                    targets: '.animejs__logo__github',
                                    color: 'hsl(0, 0%, 80%)',
                                    scale: 0.6,
                                })

                            }
                        "class=animejs__logo__github name=logo-github style=width:100%;height:100%></ion-icon> </a><a :class="$store.is_mobile ? 'is-hidden' : ''"@click.prevent="
                            () => {
                                anime({
                                    targets: '.animejs__arrow__back',
                                    rotate: button_turned ? 0 : 180,
                                });
                                button_turned = !button_turned;
                            }
                        "@mouseenter="
                        () => {
                            anime({
                                targets: '.animejs__arrow__back',
                                color: '#e50000',
                            });
                            anime({
                                targets: '.animejs__arrow__button',
                                scale: 1.2,
                            });
                        }
                    "@mouseleave="
                        () => {
                            anime({
                                targets: '.animejs__arrow__back',
                                color: 'hsl(0, 0%, 80%)',
                            });
                            anime({
                                targets: '.animejs__arrow__button',
                                scale: 1
                            });
                        }
                    "class="is-rounded is-dark animejs__arrow__button is-clickable"style=z-index:1000000> <ion-icon x-init="
                            () => {
                                anime({
                                    targets: '.animejs__arrow__back',
                                    easing: 'linear',
                                    duration: 100,
                                    color: 'hsl(0, 0%, 80%)',
                                    scale: 0.6,
                                });
                            }
                        "class=animejs__arrow__back name=arrow-back-outline style=width:100%;height:100%></ion-icon> </a></div><template x-if=!$store.is_mobile>{% if request.user.is_authenticated %} <figure class="image is-48x48 pt-2 pl-2"><img src="{% static 'core/images/placeholder.png' %}"class=is-rounded></figure> {% else %} <div class=navbar-item><div class=buttons><button @click.prevent="
                                () => {
                                    window.location = '/authentication/login/?next={{ request.path }}'
                                }
                            "class="button is-ghost is-rounded">Log in</button></div></div> {% endif %}</template></div></div></nav>
```

See the problem? ( AlpineJs is very popular with Django by the way )

>You should inspect the Accept-Encoding header to determine what compression algorithm the client supports, and use the best one available.

Yep i am doing that in [here ](https://github.com/baseplate-admin/django_strip_whitespace/blob/main/strip_whitespace/middlewares/functions/add_headers.py) but the approach is kinda flawed. But i wanted to give the developer absolute control over their choice. @adamchainz  thanks for giving me another way to approach it.

> I also think the compression middleware is best separate from the HTML minification middleware.

Okay ? I actually used this with GZIP middleware at first and thats why i wrote it like that. 

### Essentially this is my very first large project. So i made a lot of mistakes. and it would mean a lot if someone points me in the right direction

**Comment by adamchainz** [comment]
> You should look into this. Note that python regex replacing is faster and more performant than doing what minify-html...

Performance at the cost of correctness is *not* worth it. The regex based approach simply chokes on HTML *about* e.g. Alpine. For example here one of your regexes will capture from </pre> onwards:

```
For initial code use <pre>x-init="</pre> and type the appropriate code, ending with <pre>"</pre>.
```

Also I doubt that running a bunch of Python functions is indeed faster. I would like to see a benchmark. I found that minify-html can minify 10k of HTML in 14 microseconds, which is barely time to make a single Python function call.

> See the problem? ( AlpineJs is very popular with Django by the way ). I opened an Issue in upstream

Great, I added a comment to clarify what your request is.

> see the problem.

You still didn't communicate the problem in a single statement: that some bytes could still be saved. IMO this is not the end of the world.

> Essentially this is my very first large project. At the time of writing I didn't know much about good code. So i made a ton of mistakes and it would mean a lot if someone points me in the right direction. I will happily modify my module to suit the audience at large.

I appreciate this. I think you've done very well for the effort you've put in. I would like to encourage you to keep on learning. There are plenty of resources out there. I'm afraid I don't have time to mentor you.

I'm going to keep my package as-is. It's minimal and matches my tastes. OSS isn't a "competition", several solutions can exist :)

My main feedback for your package is, again, that I would always avoid regex-based HTML parsing. Try using `html.parser`!

Good luck!
